### PR TITLE
Pipelined vif feature computation: memory requirements reduced and pe…

### PR DESCRIPTION
**Pull request by foosoftsrl**
…rformance improvement

Here I am.

See https://github.com/Netflix/vmaf/pull/943 for context

Have fun

@Kyle: Had a look at the differences as seen from github. Yes the space only modifications are annoying.

### CODE REVIEW

**Comment by foosoftsrl** [comment]
I'm on my master branch, now... Not smart enough to work with branches...
regression test should be ok.

**Comment by kylophone** [comment]
Looks like the AVX-512 build is broken for me.

```sh
../src/feature/x86/vif_avx512.c:1479:31: error: member reference base type '__m256i' (vector of 4 'long long' values) is not a structure or union                                                                             
                assert(msigma1.m256i_u64[xx] == sigma1_sq);                                                                                                                                                                   
                       ~~~~~~~^~~~~~~~~~                                                                                                                                                                                      
/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/assert.h:93:25: note: expanded from macro 'assert'                                                                                                       
    (__builtin_expect(!(e), 0) ? __assert_rtn(__func__, __FILE__, __LINE__, #e) : (void)0)                                                                                                                                    
                        ^                                                                                                                                                                                                     
../src/feature/x86/vif_avx512.c:1480:32: error: member reference base type '__m256i' (vector of 4 'long long' values) is not a structure or union                                                                             
                assert(msigma12.m256i_u64[xx] == sigma12);                                                                                                                                                                    
                       ~~~~~~~~^~~~~~~~~~                                                                                                                                                                                     
/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/assert.h:93:25: note: expanded from macro 'assert'                                                                                                       
    (__builtin_expect(!(e), 0) ? __assert_rtn(__func__, __FILE__, __LINE__, #e) : (void)0)                                                                                                                                    
                        ^                                                                                                                                                                                                     
../src/feature/x86/vif_avx512.c:1481:31: error: member reference base type '__m256i' (vector of 4 'long long' values) is not a structure or union                                                                             
                assert(msigma2.m256i_u64[xx] == sigma2_sq);                                                                                                                                                                   
                       ~~~~~~~^~~~~~~~~~                                                                                                                                                                                      
/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/assert.h:93:25: note: expanded from macro 'assert'                                                                                                       
    (__builtin_expect(!(e), 0) ? __assert_rtn(__func__, __FILE__, __LINE__, #e) : (void)0)                                                                                                                                    
                        ^                                                                                                                                                                                                     
../src/feature/x86/vif_avx512.c:1487:30: error: member reference base type '__m256i' (vector of 4 'long long' values) is not a structure or union                                                                             
                    assert(mx.m256i_u64[xx] == x);                                                                                                                                                                            
                           ~~^~~~~~~~~~
/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/assert.h:93:25: note: expanded from macro 'assert'
    (__builtin_expect(!(e), 0) ? __assert_rtn(__func__, __FILE__, __LINE__, #e) : (void)0)
                        ^
../src/feature/x86/vif_avx512.c:1488:37: error: member reference base type '__m256i' (vector of 4 'long long' values) is not a structure or union
                    assert(mlog_den1.m256i_u64[xx] == log_den1);
                           ~~~~~~~~~^~~~~~~~~~
/Library/Developer/CommandLineTools/SDKs/MacOSX10.15.sdk/usr/include/assert.h:93:25: note: expanded from macro 'assert'
    (__builtin_expect(!(e), 0) ? __assert_rtn(__func__, __FILE__, __LINE__, #e) : (void)0)
                        ^
../src/feature/x86/vif_avx512.c:1502:36: error: member reference base type '__m256i' (vector of 4 'long long' values) is not a structure or union
                    assert(maccum_x.m256i_u64[xx] == accum_x[xx]);
                           ~~~~~~~~^~~~~~~~~~
```

**Comment by kylophone** [comment]
Alright, here is some quick and dirty benchmarking from me.

- **C**: -5.7% runtime. -31% memory. 
- **AVX2**: -18.6% runtime. -34.2% memory.
- **AVX-512**: untested, broken build.

```sh
+ /usr/bin/time -lp ./build/tools/vmaf --cpumask -1 --reference Sniper_1920x1080P_30fps_8bit.y4m --distorted Sniper_1920x1080P_30fps_8bit.y4m --output /dev/null
real        33.15
user        32.84
sys          0.19
 133316608  maximum resident set size
         0  average shared memory size
         0  average unshared data size
         0  average unshared stack size
     32569  page reclaims
         0  page faults
         0  swaps
         0  block input operations
         0  block output operations
         0  messages sent
         0  messages received
         0  signals received
        33  voluntary context switches
      1010  involuntary context switches
+ /usr/bin/time -lp ./build_pr/tools/vmaf --cpumask -1 --reference Sniper_1920x1080P_30fps_8bit.y4m --distorted Sniper_1920x1080P_30fps_8bit.y4m --output /dev/null
real        31.27
user        31.12
sys          0.13
  91754496  maximum resident set size
         0  average shared memory size
         0  average unshared data size
         0  average unshared stack size
     22422  page reclaims
         0  page faults
         0  swaps
         0  block input operations
         0  block output operations
         0  messages sent
         0  messages received
         0  signals received
         0  voluntary context switches
       511  involuntary context switches
+ /usr/bin/time -lp ./build/tools/vmaf --reference Sniper_1920x1080P_30fps_8bit.y4m --distorted Sniper_1920x1080P_30fps_8bit.y4m --output /dev/null
real        13.27
user        13.12
sys          0.14
 133287936  maximum resident set size
         0  average shared memory size
         0  average unshared data size
         0  average unshared stack size
     32562  page reclaims
         0  page faults
         0  swaps
         0  block input operations
         0  block output operations
         0  messages sent
         0  messages received
         0  signals received
         0  voluntary context switches
       301  involuntary context switches
+ /usr/bin/time -lp ./build_pr/tools/vmaf --reference Sniper_1920x1080P_30fps_8bit.y4m --distorted Sniper_1920x1080P_30fps_8bit.y4m --output /dev/null
real        10.81
user        10.67
sys          0.13
  87646208  maximum resident set size
         0  average shared memory size
         0  average unshared data size
         0  average unshared stack size
     21419  page reclaims
         0  page faults
         0  swaps
         0  block input operations
         0  block output operations
         0  messages sent
         0  messages received
         0  signals received
         0  voluntary context switches
       193  involuntary context switches
```

**Comment by kylophone** [comment]
Alright, I think this is great work! The speedups are really impressive, and the pipelining technique is certainly the way to go for VIF. We had talked about using this type of technique last year when we did the original optimizations, but never got around to it. So, thank you for  Once we address the broken AVX-512 build, I think we can start cleaning this up and getting it shape for merging. There is a high bar to get code like this into libvmaf, so hopefully you don't feel like I'm picking on you especially. Please be patient since this is a _HUGE_ PR, and it's a holiday week here in the US. It may take a while.

- Code is numerically correct for all instruction sets and passes all regression tests.
- There is also an internal set of sequences we have that I can run this on. I will share any mismatches we find.
- Code should be platform agnostic, portable, and cross platform.
- Commits should be focused, clean and free of distracting whitespace changes, re-indentations, etc.
- ...

Once we fix the AVX-512 build, we can start addressing individual concerns. I am happy to either help out with the code directly, or just provide inline comments.

**Comment by foosoftsrl** [comment]
Hi Kyle,
yes contributing is a bit frustrating, and yes I understand that there are good reasons for you to be picky.
Let's keep this thing going.
I'll fix the problem with AVX-512 shortly... there is a sort of "experimental" code which will be a big speed booster for AVX 512 which I had no time / confidence to actually "deploy".
Support in form of inline comments / code is more than appreciated: it is needed :)

**Comment by foosoftsrl** [comment]
Hi Kyle,
did dome cleanup, avx512 should build.
Huge optimization for avx512 of the vif computation (I mean... that part with logarithms).
Some cleanup here and there.
As main concern (AVX-512 build) is addressed, we should now be able to continue!

**Comment by foosoftsrl** [comment]
PS. the AVX512 code for VIF computation is to be checked carefully

**Comment by kylophone** [comment]
```
../src/feature/x86/vif_avx512.c:115:65: error: always_inline function '_mm512_lzcnt_epi64' requires target feature 'avx512cd', but would be inlined into function 'vif_statistic_avx512' that is compiled without support for 'avx512cd'
        __m512i mnorm = _mm512_sub_epi64(_mm512_set1_epi64(48), _mm512_lzcnt_epi64(mlog_den_stage1));
                                                                ^
../src/feature/x86/vif_avx512.c:133:70: error: always_inline function '_mm512_lzcnt_epi64' requires target feature 'avx512cd', but would be inlined into function 'vif_statistic_avx512' that is compiled without support for 'avx512cd'
        __m512i mnumer1_lz = _mm512_sub_epi64(_mm512_set1_epi64(48), _mm512_lzcnt_epi64(mnumer1));
                                                                     ^
../src/feature/x86/vif_avx512.c:139:74: error: always_inline function '_mm512_lzcnt_epi64' requires target feature 'avx512cd', but would be inlined into function 'vif_statistic_avx512' that is compiled without support for 'avx512cd'
        __m512i mnumer1_tmp_lz = _mm512_sub_epi64(_mm512_set1_epi64(48), _mm512_lzcnt_epi64(mnumer1_tmp));
```

File: libvmaf/src/feature/integer_vif.c [hunk 1]
```diff
@@ -16,10 +16,13 @@
  *
  */
 
+#include <immintrin.h>
```
Comment (kylophone) [inline libvmaf/src/feature/integer_vif.c:19]: The C implementation needs to be multiplatform. This is an x86-only header.

File: libvmaf/src/feature/integer_vif.c [hunk 2]
```diff
@@ -16,10 +16,13 @@
...
+
 #include <errno.h>
 #include <math.h>
 #include <string.h>
 #include <stdio.h>
+#include <assert.h>
```
Comment (kylophone) [inline libvmaf/src/feature/integer_vif.c:25]: Probably should avoid asserts in the final code.

File: libvmaf/src/feature/integer_vif.c [hunk 3]
```diff
@@ -37,758 +40,800 @@
...
-        .name = "vif_enhn_gain_limit",
-        .help = "enhancement gain imposed on vif, must be >= 1.0, "
-                "where 1.0 means the gain is completely disabled",
-        .offset = offsetof(VifState, vif_enhn_gain_limit),
-        .type = VMAF_OPT_TYPE_DOUBLE,
-        .default_val.d = DEFAULT_VIF_ENHN_GAIN_LIMIT,
-        .min = 1.0,
-        .max = DEFAULT_VIF_ENHN_GAIN_LIMIT,
-    },
-    { 0 }
```
Comment (kylophone) [inline libvmaf/src/feature/integer_vif.c:71]: This is an example of a whitespace only change.  It shows up in the diff, but nothing is changed. Avoid these, they're all over the place.

**Comment by foosoftsrl** [comment]
**AVX512 compilation error**

it's ok in GCC... I was a bit lazy and did not put the needed avx512 flags in the meson build. 
A github task would make it easier to spot these kind of things.
I can try and patch the code, but as you may have guessed, I changed the processor target, requiring more bits of AVX512.
Is it ok? At a bare minimum, the avx512 detection code must be modified.
Here I'm on 11-th generation consumer processor. No problems.
I read somewhere that AVX512 is all-or-none in most processors.

**immintrin.h inclusion**

typo

**assert.h**

if you say they should "probably" be removed, I'd prefer to leave them there. They're a life-saver, and can't figure out how to have two code bases, one asserted, the other not.
I'm open to suggestions, seriously

**whitespace only changes**

I'll try to fix some of those quickly

I assume that for the moment you won't touch the code (not sure you have rights) so I'll try and fix on my own rapidly.

**Comment by foosoftsrl** [comment]
Hi,
the assert is still there. I'd prefer to leave it there until things settle down.
the other points are more or less ok.
The whitespace issue... I did my best.
Perhaps macros or inline functions could make the code more readable and comparable without not too much impact

**Comment by foosoftsrl** [comment]
Any news?

**Comment by kylophone** [comment]
> Any news?

AVX-512 builds fine now. I am still verifying the results. It's a huge PR, si it will take me a little bit more time to get through it. I will update you soon.

**Pull request closed without merging**